### PR TITLE
Rework Annotations UI + annotation class refactor

### PR DESCRIPTION
**Pull request by cleaaum**
This PR re works the UI for the annotation class and annotation tools. It also refactors the way the annotation classes are defined in memory. They are implemented using an OOP approach in which each class has its own set of properties such as as visibility, annotations, color, label etc. These can now easily be processed using MATCH callbacks. 
image

### CODE REVIEW

File: components/annotation_class.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+import dash_mantine_components as dmc
+from dash import html, dcc
+from dash_iconify import DashIconify
+
+
+# This function generates a class component with all its buttons (hide/show, edit, delete)
+# The class color is used as an ID to identify a class (as all class colors are unique and cannot be modified)
+def annotation_class_item(class_color, class_label):
```
Comment (hannahker) [inline components/annotation_class.py:8]: Could we add `cursor: 'pointer'` to the style for these items so that they are clearly clickable?

Comment (cleaaum) [reply components/annotation_class.py:8]: Yes I added a hover as well as a curser and styled the border and background color a bit more.



https://github.com/mlexchange/mlex_highres_segmentation/assets/56569375/c33801de-bd29-4743-8462-68144905243e

Comment (hannahker) [reply components/annotation_class.py:8]: Love!

**Comment by hannahker** [comment]
Closes #75 and #94

File: callbacks/image_viewer.py [hunk 2]
```diff
@@ -262,27 +238,59 @@ def update_viewfinder(relayout_data, annotation_store):
...
+def locally_store_annotations(
+    relayout_data, img_idx, annotation_store, all_annotation_class_store, current_color
+):
     """
-    Upon finishing relayout event (drawing, but it also includes panning, zooming),
-    this function takes the annotation shapes, and stores it in the dcc.Store, which is then used elsewhere
-    to preserve drawn annotations, or to save them.
+    Upon finishing a relayout event (drawing, panning or zooming), this function takes the
+    currently drawn shapes or zoom/pan data, and stores the lastest added shape to the appropriate class-annotation-store,
+    or the image pan/zzom position to the anntations-store.
```
Comment (hannahker) [inline callbacks/image_viewer.py:258]: Some minor typos here

File: callbacks/image_viewer.py [hunk 3]
```diff
@@ -262,27 +238,59 @@ def update_viewfinder(relayout_data, annotation_store):
...
+
+@callback(
+    Output(
+        {"type": "annotation-class-store", "index": ALL}, "data", allow_duplicate=True
+    ),
+    Input("project-name-src", "value"),
+    State({"type": "annotation-class-store", "index": ALL}, "data"),
+    prevent_initial_call=True,
+)
+def clear_annotations_on_dataset_cahnge(change_project, all_annotation_class_store):
```
Comment (hannahker) [inline callbacks/image_viewer.py:286]: typo: `cahnge` vs `change`

File: callbacks/control_bar.py [hunk 4]
```diff
@@ -39,6 +39,60 @@
...
+    return current_selection
+
+
+@callback(
+    Output({"type": "annotation-class", "index": ALL}, "style"),
+    Input("current-class-selection", "data"),
+    State({"type": "annotation-class", "index": ALL}, "id"),
+)
+def update_selected_class_style(selected_class, current_ids):
+    """This callback is responsible for updating the style of the selected annot class to makw it appear
```
Comment (hannahker) [inline callbacks/control_bar.py:65]: Typo

File: callbacks/control_bar.py [hunk 5]
```diff
@@ -213,465 +242,263 @@ def annotation_width(width_value):
...
-    class_label,
-    class_color,
-    current_classes,
-    classes_to_delete,
-    classes_to_hide,
-    old_label,
-    new_label,
-    annotation_store,
-    image_idx,
+def delete_annotation_class(
```
Comment (hannahker) [inline callbacks/control_bar.py:474]: Wow look at that simplification 😍

**Comment by hannahker** [comment]
This looks amazing! I've spent time clicking around the UI and can't seem to be able to break anything. This is in the stage of fixing typos and nit-picky UI tweaks. 

- Can we update the styling on this disclaimer text so that it's more aligned with the rest of the UI? (ie. update the font and add some padding)

Screenshot 2023-09-06 at 2 10 01 PM

- Can we reduce the amount of white space between the "Annotation tools" header and the "Drawing toolbar" text? 
- Can we expand the width of the drawing toolbar so that it matches the full width of each class button?
- We should add some padding to the right of the garbage icon so that it doesn't come right up to the edge of the button when it's highlighted. 

Screenshot 2023-09-06 at 2 08 40 PM


Otherwise, this should be good to merge once I finish up with the work to update the annotations data handling.

**Review by hannahker — APPROVED** [review]
Amazing work!! 💃 💃

**Pull request merged**